### PR TITLE
Workaround for Dataframe object in the config: check whether value is string before mandatory missing check

### LINKED ISSUE #566: Illegal Comparisons when handling pandas DataFrame objects
**Describe the bug**
In several places, the code attempts to check if a node is missing with a comparison + a conditional (eg `x = "a" if node == "???" else "b"`). This conditional breaks if the node value is a pandas DataFrame, as DataFrames implement additional `__bool__` logic that prevents such a conditional (see: https://github.com/pandas-dev/pandas/blob/v1.2.2/pandas/core/generic.py#L1440-L1446).

**To Reproduce**
creating a DictConfig:
```
from pandas import DataFrame
from omegaconf import OmegaConf

kwargs = OmegaConf.create({"a": DataFrame()}, flags={"allow_objects": True})
```

when attempting to set an attribute of an existing config:
```
from pandas import DataFrame
from omegaconf import OmegaConf

c = OmegaConf.create()
c._set_flag("allow_objects", True)
c.a = DataFrame()
```

**Expected behavior**
when running creating DictConfig or setting a new attr, there should simply be an update to the config where the DF object is now included (assuming allow_object flag is set to True).

**Additional context**
- [X] OmegaConf version: master branch
- [X] Python version: 3.7.6
- [X] Operating system : ubuntu
- [X] Please provide a minimal repro

### PR DESCRIPTION
**Pull request by jeffknaide**
Closes #566 

sidestep source of https://github.com/facebookresearch/hydra/issues/1285 at the omegaconf level

### CODE REVIEW

File: omegaconf/basecontainer.py [hunk 1]
```diff
@@ -784,7 +784,7 @@ def _create_structured_with_missing_fields(
 def _is_missing_value(value: Any) -> bool:
     if isinstance(value, Container):
         value = value._value()
-    ret = value == "???"
+    ret = value == "???" if isinstance(value, str) else False
```
Comment (Jasha10) [inline omegaconf/basecontainer.py:787]: So the idea is to sidestep the custom `DataFrame.__eq__` method by first deciding if `value` is a `str`, then only calling `value.__eq__("???")` if we are sure that `value` is a `str`.

Comment (jeffknaide) [reply omegaconf/basecontainer.py:787]: yep, exactly.

**Comment by omry** [comment]
Thanks @jeffknaide.
We have some outstanding discussions around the internal behavior of MISSING, we will address this once we comes to a conclusion there.

**Comment by omry** [comment]
Blocked by #543.

File: omegaconf/_utils.py [hunk 2]
```diff
@@ -353,6 +353,9 @@ def get_value_kind(
 
     value = _get_value(value)
 
+    if not isinstance(value, str):
+        return ValueKind.VALUE
+
```
Comment (omry) [inline omegaconf/_utils.py:358]: Change the following check to:
```
    if isinstance(value,str) and value == "???":
        return ValueKind.MANDATORY_MISSING
```
instead.

File: omegaconf/basecontainer.py (hunk 1 above)
Comment (omry) [inline omegaconf/basecontainer.py:787]:
```suggestion
    ret = isinstance(value, str) and value == "???"
```

File: tests/__init__.py [hunk 3]
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
         raise NotImplementedError()
 
 
+class Dataframe:
+    def __init__(self) -> None:
+        pass
+
+    def __eq__(self, other: Any) -> Any:
+        return self
```
Comment (omry) [inline tests/__init__.py:42]: I find `__eq__` returning self very odd. what's going on here?

File: tests/__init__.py [hunk 4]
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
         raise NotImplementedError()
 
 
+class Dataframe:
```
Comment (omry) [inline tests/__init__.py:37]: Add a comment here.
something like:

```
Emulates Pandas Dataframe equality behavior.
```

File: tests/__init__.py [hunk 5]
```diff
@@ -34,6 +34,17 @@ def __deepcopy__(self, memo: Any) -> Any:
...
+    def __init__(self) -> None:
+        pass
+
+    def __eq__(self, other: Any) -> Any:
+        return self
+
+    def __bool__(self) -> None:
+        raise ValueError
```
Comment (omry) [inline tests/__init__.py:45]: add a comment explaining what this achieves.

File: tests/test_merge.py [hunk 6]
```diff
@@ -599,6 +600,14 @@ def test_merge_allow_objects(merge: Any) -> None:
     assert ret == {"a": 10, "foo": iv}
 
 
+def test_merge_with_allow_Dataframe() -> None:
+    kwargs = OmegaConf.create({"a": Dataframe()}, flags={"allow_objects": True})
+    cfg = OmegaConf.create()
+
+    cfg.merge_with(kwargs)
+    assert isinstance(cfg.a, Dataframe)
```
Comment (omry) [inline tests/test_merge.py:608]: If this still reproduces the problem, please switch to it:
```suggestion
def test_merge_with_allow_Dataframe() -> None:
    cfg = OmegaConf.create({"a": Dataframe()}, flags={"allow_objects": True})
    ret = OmegaConf.merge({}, cfg)
    assert isinstance(ret.a, Dataframe)
```

File: tests/test_nodes.py [hunk 7]
```diff
@@ -559,3 +559,10 @@ def test_allow_objects() -> None:
     iv = IllegalType()
     c.foo = iv
     assert c.foo == iv
+
+
+def test_allow_Dataframe_objects() -> None:
+    c = OmegaConf.create()
+    c._set_flag("allow_objects", True)
+    c.a = Dataframe()
+    assert isinstance(c.a, Dataframe)
```
Comment (omry) [inline tests/test_nodes.py:568]: allow_objects is already tested, remove this test.

**Review by omry — CHANGES_REQUESTED** [review]
Thanks @jeffknaide.
See inline comments.

File: tests/__init__.py (hunk 3 above)
Comment (jeffknaide) [reply tests/__init__.py:42]: when a pandas DataFrame runs `df.__eq__(val)`, it returns a DataFrame of booleans with the same shape as the original DataFrame. Here I was trying to mimic the behavior of a df comparison resulting in the same class being returned, as this is partially what caused issues originally. perhaps this should be made clearer with a comment?

Comment (omry) [reply tests/__init__.py:42]: Gotcha.
Yes, add a comment what the behavior we are mimicking is.

Comment (jeffknaide) [reply tests/__init__.py:42]: i don't think so -- if we don't override eq, the comparison to string would return a bool and then not raise the expected error on the conditional. eg `if Dataframe() == "???"` would equate to `if False` (which wouldn't raise the expected error) without the eq override.

File: tests/__init__.py (hunk 5 above)
Comment (jeffknaide) [reply tests/__init__.py:49]: let me know if the comments I added here work

**Comment by omry** [comment]
lgtm.
@odelalleau, this could interact with your outstanding PR to change the semantics of missing.

**Comment by odelalleau** [comment]
> lgtm.
> @odelalleau, this could interact with your outstanding PR to change the semantics of missing.

At first glance I don't foresee any issue (*crossing fingers*). Just mentioning that maybe there could be a new utility function to avoid repeating `if isinstance(x, str) and x == "???"` everywhere?

**Comment by omry** [comment]
Yeah, was thinking that too.

**Comment by jeffknaide** [comment]
@odelalleau @omry oh, yes, that makes sense. I just added `is_mandatory_missing` to `_utils` and referenced it where i had been using `if isinstance(x, str) and x == "???"`.

There was also an inner function defined in `BaseContainer._resolve_with_default` called `is_mandatory_missing` which was returned of `get_value_kind`. I just removed that inner function in favor of `is_mandatory_missing` defined in _utils. This seems to break the tests, though, so I must be missing something that's happening within `get_value_kind`

**Comment by odelalleau** [comment]
> I must be missing something that's happening within `get_value_kind`

It’s almost certainly because that function can be called on a node: it starts with `_get_value(..)` to obtain the node’s value if that’s the case. So in BaseContainer the problem is likely related to ˋ if is_mandatory_missing(resolved):`, and replacing ˋresolved` with its value here should fix it.

**Comment by omry** [comment]
I think we will want to take a closer look at this in a followup.
can you undo the refactoring changes?

**Comment by odelalleau** [comment]
> can you undo the refactoring changes?

Most could be kept, only the change in `_resolve_with_default()` needs reverting

**Comment by omry** [comment]
Maybe. it just feel like there are a bit too many layers here. The logic can be slightly reorganized to be simpler.

**Comment by jeffknaide** [comment]
@omry - i just reverted previous commit. let me know if you want me to put the utils function back while leaving `_resolve_with_default ` alone.

**Review by omry — APPROVED** [review]
Thanks @jeffknaide.

**Pull request merged**